### PR TITLE
add object lock configuration APIs

### PR DESCRIPTION
**Pull request by balamurugana**

### CODE REVIEW

File: docs/API.md [hunk 1]
```diff
@@ -1565,6 +1565,75 @@ if err != nil {
...
+|:---|:---| :---|
+|`mode` | _RetentionMode_  |Current retention mode |
+|`validity` | _uint_  |Current validity period |
+|`unit` | _ValidityUnit_  |Unit of validity period |
+|`err` | _error_  |Standard Error  |
+
+__Example__
+
+```go
+mode, validity, unit, err := minioClient.GetBucketLifecycle("my-bucketname")
```
Comment (sinhaashish) [inline docs/API.md]: `GetBucketLifecycle` -> `GetBucketObjectLockConfig`
`minioClient`-> `s3Client` as it is an s3 feature till now.

Comment (kannappanr) [reply docs/API.md]: @balamurugana Can you please fix the typo here? Should we add versioning in `makeBucket` as we have done in `minio-java`?

Comment (balamurugana) [reply docs/API.md]: Fixed the doc.

No. it breaks backward compatibility.  I added 
```go
EnableVersioning(bucketName string) error
DisableVersioning(bucketName string) error
```

File: examples/s3/getbucketobjectlockconfig.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	// determined based on the Endpoint value.
+	s3Client, err := minio.New("s3.amazonaws.com", "YOUR-ACCESSKEYID", "YOUR-SECRETACCESSKEY", true)
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	// s3Client.TraceOn(os.Stderr)
+
+	// Get object lock configuration.
+	mode, validity, unit, err := minioClient.GetBucketLifecycle("my-bucketname")
```
Comment (kannappanr) [inline examples/s3/getbucketobjectlockconfig.go]: Typo should be fixed.

File: examples/s3/getbucketobjectlockconfig.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	// determined based on the Endpoint value.
+	s3Client, err := minio.New("s3.amazonaws.com", "YOUR-ACCESSKEYID", "YOUR-SECRETACCESSKEY", true)
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	// s3Client.TraceOn(os.Stderr)
+
+	// Get object lock configuration.
+	mode, validity, unit, err := minioClient.GetBucketObjectLockConfig("my-bucketname")
```
Comment (kannappanr) [inline examples/s3/getbucketobjectlockconfig.go]:
```suggestion
	mode, validity, unit, err := s3Client.GetBucketObjectLockConfig("my-bucketname")
```

File: examples/s3/setbucketobjectlockconfig.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+	}
+
+	// s3Client.TraceOn(os.Stderr)
+
+	// Set object lock configuration.
+	mode := Governance
+	validity := uint(30)
+	unit := Days
+
+	err = minioClient.SetBucketObjectLockConfig("my-bucketname", &mode, &validity, &unit)
```
Comment (kannappanr) [inline examples/s3/setbucketobjectlockconfig.go]:
```suggestion
	err = s3Client.SetBucketObjectLockConfig("my-bucketname", &mode, &validity, &unit)
```

**Comment by sinhaashish** [comment]
In minio-java we have introduced a new api `makeBucket(String bucketName, String region, boolean objectLock)` .If the objectLock is set to true then the bucket can support locking mechanism.. For this we pass `x-amz-bucket-object-lock-enabled` as true in the header.

This functionality is not present in `minio-go` we need to introduce new make bucket api which supports locking functionality. 

As of now if we try to set the lock we get a message stating ` Object Lock configuration cannot be enabled on existing buckets` as the bucket was not lock enabled. 

This complete testing of this PR will be done; only when bucket is lock enabled and we can get/set the lock configuration. @kannappanr  Please confirm on adding a new api for `makeBucket`

**Comment by kannappanr** [comment]
@balamurugana Please add a new API for `MakeBucket`. It will be tentatively named `MakeObjectLockEnabledBucket`

**Comment by balamurugana** [comment]
@kannappanr 

> @balamurugana Please add a new API for `MakeBucket`. It will be tentatively named `MakeObjectLockEnabledBucket`

Please confirm the name officially. Adding the function is just one liner.

**Comment by kannappanr** [comment]
@sinhaashish can PTAL?

**Comment by kannappanr** [comment]
@sinhaashish @krishnasrinivas PTAL

File: api-put-bucket.go [hunk 5]
```diff
@@ -98,6 +97,28 @@ func (c Client) MakeBucket(bucketName string, location string) (err error) {
...
+}
+
+// MakeBucketWithObjectLock creates a object lock enabled new bucket with bucketName.
+//
+// Location is an optional argument, by default all buckets are
+// created in US Standard Region.
+//
+// For Amazon S3 for more supported regions - http://docs.aws.amazon.com/general/latest/gr/rande.html
+// For Google Cloud Storage for more supported regions - https://cloud.google.com/storage/docs/bucket-locations
+func (c Client) MakeBucketWithObjectLock(bucketName string, location string) (err error) {
```
Comment (sinhaashish) [inline api-put-bucket.go]: IMO we should have `MakeBucketWithObjectLock` in examples . 
But for sure it must be added in the docs.

File: examples/s3/setbucketobjectlockconfig.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+	// s3Client.TraceOn(os.Stderr)
+
+	// Set object lock configuration.
+	mode := minio.Governance
+	validity := uint(30)
+	unit := minio.Days
+
+	err = s3Client.SetBucketObjectLockConfig("my-bucketname", &mode, &validity, &unit)
+	if err != nil {
```
Comment (sinhaashish) [inline examples/s3/setbucketobjectlockconfig.go]: We can add a print for the success scenerio like `"Object lock configuration set successfully"`

File: examples/s3/getbucketobjectlockconfig.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+	// s3Client.TraceOn(os.Stderr)
+
+	// Get object lock configuration.
+	mode, validity, unit, err := s3Client.GetBucketObjectLockConfig("my-bucketname")
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	if mode != nil {
+		fmt.Printf("%v mode is enabled for %v %v for bucket 'my-bucketname'\n", mode, validity, unit)
```
Comment (sinhaashish) [inline examples/s3/getbucketobjectlockconfig.go]: After running this file got the below o/p
`COMPLIANCE mode is enabled for 0xc0001d2130 DAYS for bucket 'my-bucketname'
`
The value for validity is not shown correctly.

Comment (kannappanr) [inline examples/s3/getbucketobjectlockconfig.go]:
```suggestion
		fmt.Printf("%v mode is enabled for %v %v for bucket 'my-bucketname'\n", mode, *validity, unit)
```

File: api-put-bucket.go (hunk 5 above)
Comment (balamurugana) [reply api-put-bucket.go]: all done

File: examples/s3/setbucketobjectlockconfig.go (hunk 6 above)
Comment (balamurugana) [reply examples/s3/setbucketobjectlockconfig.go]: done

File: examples/s3/getbucketobjectlockconfig.go (hunk 7 above)
Comment (balamurugana) [reply examples/s3/getbucketobjectlockconfig.go]: fixed

**Pull request merged**